### DISCUSSION TITLE
Getting Segment Values while looping in a for each loop

### DISCUSSION THREAD

**Discussion by Kuepp1** [nextgenhealthcare/connect#5880] [General]
Hello guys,

i'm having some ORU which contain one OBR and many OBX. For each OBX i need to create one additional OBR segment.
I also need to fill a few things into the new created segment like a timestamp from MSH7

Sample OBX:
OBX|3|ST|^FIO2^I|400063|21.0|%||N|||F|||||||||||

I'm trying it like this:

`var ts = msg['MSH']['MSH.7']['MSH.7.1'].toString();

for each (seg in msg.children()) {
	if (seg.name().toString() == "OBX") {
		var newSeg = createSegment('OBR');
		//Set TIMESTAMP
		newSeg['OBR.6']['OBR.6.1'] = ts;
		insertSegmentBefore(newSeg, seg);
	}
}`

So far that's doing its job. However i need to also add in the value my OBX contains in 4.1.
I was hoping i could the value simply through

`var ts = msg['MSH']['MSH.7']['MSH.7.1'].toString();

for each (seg in msg.children()) {
	if (seg.name().toString() == "OBX") {
		var newSeg = createSegment('OBR');
                // add code from OBX
		newSeg['OBR.4']['OBR.4.1'] = 	seg['OBX.4']['OBX.4.1'].toString();
		//Set TIMESTAMP
		newSeg['OBR.6']['OBR.6.1'] = ts;
		insertSegmentBefore(newSeg, seg);
	}
}`

Sadly this doesn't work.
How can i access the value of each OBX in that loop?

Thank you so much for help!

**Comment by pacmano1** [comment]
Post a before and after example hl7 message please.

**Reply by Kuepp1** [reply]
Alright i bet you knew it would work( :D)  because it does now and i don't get what i did wrong before.

It looks like this now
OBR||||400067||20230711095425

Sorry for soaking up your time and thank you for a lightspeed response!

**Reply by pacmano1** [reply]
I was still curious about the before and after only because I am not sure I would have done it your way.

**Reply by Kuepp1** [reply]
I'm very open to how to make it better. I just took over a job where i am maintaining our labs mirth server. I find that thing really interesting.

How would you have approached it?

**Reply by pacmano1** [reply]
Please use code blocks and format your code.  

**Again**, post a before and after example since I don't understand how your question was phrased.

If it is just create an OBR after each OBX and ignoring the fact I did not set the set ID:

Before:

```
MSH|^~\&|||||19320604212615||ORU^R01|MSG0000000001|P|2.5
PID|1||328128260||EMPASIS^KIMBERLEY^G|DUNCAN^KATIE^G||U~U|ILLOVSKY^ROBERT^I||||||||||||||||||||||||||||xxTESTxx
PV1|1|B
OBR|1|||||||||||||||||||||||||||||||||||||||||U
TQ1|1
OBX|1||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBX|2||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBX|3||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBX|4||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBX|5||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBX|6||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBX|7||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
SPM|1|||||||||||||||||19610929111740
```

See https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Create%20and%20Insert%20Segment%20Templates 


```JS
for each (obx in msg.OBX) {
	var newobr = createSegmentAfter('OBR', obx)
	newobr['OBR.6']['OBR.6.1'] = msg['MSH']['MSH.7']['MSH.7.1'].toString()
}
```


After:
```
MSH|^~\&|||||19320604212615||ORU^R01|MSG0000000001|P|2.5
PID|1||328128260||EMPASIS^KIMBERLEY^G|DUNCAN^KATIE^G||U~U|ILLOVSKY^ROBERT^I||||||||||||||||||||||||||||xxTESTxx
PV1|1|B
OBR|1|||||||||||||||||||||||||||||||||||||||||U
TQ1|1
OBX|1||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
OBX|2||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
OBX|3||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
OBX|4||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
OBX|5||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
OBX|6||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
OBX|7||||et dolore magna aliquyam erat, sed diam voluptua. At vero eos||||||W
OBR||||||19320604212615
SPM|1|||||||||||||||||19610929111740
```

**Reply by Kuepp1** [reply]
So sorry for a late reply.
I changed my code to iterating over the OBX only. And i was able to access the actual OBX element through: 

```
var obrIndex = 0;
//ADD new OBR's for each OBX
for each (obx in msg.OBX){
	obrIndex += 1;
	var newSeg = createSegment('OBR');
	newSeg['OBR.1']['OBR.1.1'] = obrIndex; 
	newSeg['OBR.2']['OBR.2.1'] = anr;
	newSeg['OBR.4']['OBR.4.1'] = obx['OBX.3']['OBX.3.1'].toString(); 
	insertSegmentBefore(newSeg, obx);//Insert new segment as OBR before each OBX
}
```

Thank you for your help!
